Main submit

kailiangz1 opened the pull request:
add virtio vdpa driver

paravmellanox commented on drivers/vdpa/virtio/virtio.c:
```diff
@@ -0,0 +1,45 @@
+ * Copyright (c) 2022 NVIDIA Corporation & Affiliates
+ */
+
+#include <unistd.h>
```
@kailiangz1 
Can you please take the whole code of this patch out of drivers/vdpa directory?

it should be in drivers/virtio or if its library it should be in lib/virtio.

paravmellanox commented on drivers/vdpa/virtio/virtio_rxtx.h:
```diff
@@ -0,0 +1,62 @@
+ * Copyright (c) 2022 NVIDIA Corporation & Affiliates
+ */
+
+#ifndef _VIRTIO_RXTX_H_
```
VIRTNET_RX_TX_H

paravmellanox commented on drivers/vdpa/virtio/virtio_rxtx.h:
```diff
@@ -0,0 +1,62 @@
+	uint64_t	multicast;
+	uint64_t	broadcast;
+	/* Size bins in array as RFC 2819, undersized [0], 64 [1], etc */
+	uint64_t	size_bins[8];
```
why is this needed?

paravmellanox commented on drivers/vdpa/virtio/virtio_rxtx.h:
```diff
@@ -0,0 +1,62 @@
+
+struct virtnet_rx {
+	/* dummy mbuf, for wraparound when processing RX ring. */
+	struct rte_mbuf *fake_mbuf;
```
we are not working on any mbufs here. please remove all mbuf related code.

paravmellanox commented on drivers/vdpa/virtio/virtio_rxtx.h:
```diff
@@ -0,0 +1,62 @@
+	uint64_t mbuf_initializer; /**< value to init mbufs. */
+	struct rte_mempool *mpool; /**< mempool for mbuf allocation */
+
+	uint16_t queue_id;   /**< DPDK queue index. */
```
all of this.

paravmellanox commented on drivers/vdpa/virtio/virtio_rxtx.h:
```diff
@@ -0,0 +1,62 @@
+	/* Statistics */
+	struct virtnet_stats stats;
+
+	const struct rte_memzone *mz; /**< mem zone to populate RX ring. */
```
this too.

paravmellanox commented on drivers/vdpa/virtio/virtio_rxtx.h:
```diff
@@ -0,0 +1,62 @@
+};
+
+int virtio_rxq_vec_setup(struct virtnet_rx *rxvq);
+void virtio_update_packet_stats(struct virtnet_stats *stats,
```
every single API that we add, needs to have comment section to describe what the API is and what it does.

paravmellanox commented on drivers/vdpa/virtio/virtio_ring.h:
```diff
@@ -0,0 +1,188 @@
+{
+	vr->num = num;
+	vr->desc = (struct vring_desc *) p;
+	vr->avail = (struct vring_avail *) (p +
```
please do not split the code like this in multiple lines.

paravmellanox commented on drivers/vdpa/virtio/virtio_ring.h:
```diff
@@ -0,0 +1,188 @@
+{
+	size_t size;
+
+	if (virtio_with_packed_queue(hw)) {
```
we dont support packed queue right now, please remove this code.

paravmellanox commented on drivers/vdpa/virtio/virtio_ring.h:
```diff
@@ -0,0 +1,188 @@
+	struct vring_used  *used;
+};
+
+/* The standard layout for the ring is a continuous chunk of memory which
```
remove the dead code.

paravmellanox commented on drivers/vdpa/virtio/virtio_ring.h:
```diff
@@ -0,0 +1,188 @@
+#define RING_EVENT_FLAGS_ENABLE 0x0
+#define RING_EVENT_FLAGS_DISABLE 0x1
+#define RING_EVENT_FLAGS_DESC 0x2
+struct vring_packed_desc_event {
```
need empty line before the struct

paravmellanox commented on drivers/vdpa/virtio/virtio_pci.h:
```diff
@@ -0,0 +1,146 @@
+#define VTPCI_IO(hw) (&((hw)->io))
+#define VTPCI_DEV(hw) ((hw)->pci_dev)
+
+
```
please review whole code for these empty lines.

kailiangz1 replied:
Done

paravmellanox commented on drivers/vdpa/virtio/virtio_pci.h:
```diff
@@ -0,0 +1,146 @@
+	bool modern;
+	uint8_t notify_bar;
+	int vfio_dev_fd;
+};
```
naturally pack the structure so that there are no holes.

kailiangz1 replied:
Done

paravmellanox commented on drivers/vdpa/virtio/virtio_pci.c:
```diff
@@ -0,0 +1,809 @@
+		(((dev)->msix_status == VIRTIO_MSIX_ENABLED) ? 24 : 20)
+
+static inline int
+check_vq_phys_addr_ok(struct virtqueue *vq)
```
any check function that doesnt modify the structure, must take const pointer.
and if it returns boolean 0 or 1, change it to return boolean instead of int.

paravmellanox commented on drivers/vdpa/virtio/virtio_pci.c:
```diff
@@ -0,0 +1,809 @@
+}
+
+static void
+legacy_write_dev_config(struct virtio_hw *hw, size_t offset,
```
module_submodule_object_verbs

legacy_dev_config_write()

Align the whole code.

paravmellanox commented on drivers/vdpa/virtio/virtio_api.c:
```diff
@@ -0,0 +1,321 @@
+}
+
+uint16_t
+virtio_dev_get_nr_vq(struct virtio_pci_dev *vpdev)
```
virtio_dev_nr_vq_get

kailiangz1 replied:
Done

paravmellanox commented on drivers/vdpa/virtio/virtio_api.c:
```diff
@@ -0,0 +1,321 @@
+
+	nr_vq = hw->max_queue_pairs * 2;
+	if (virtio_with_feature(hw, VIRTIO_NET_F_CTRL_VQ))
+		nr_vq += 1;
```
no. this is net specific code. it cannot be generic here.

paravmellanox commented on drivers/vdpa/virtio/virtio_api.c:
```diff
@@ -0,0 +1,321 @@
+	PMD_INIT_LOG(DEBUG,"Guest_features is 0x%"PRIx64,hw->guest_features);
+	return vpdev;
+
+	error:
```
label must start at start of the line.

kailiangz1 replied:
Done

paravmellanox commented on drivers/vdpa/virtio/virtio_api.c:
```diff
@@ -0,0 +1,321 @@
+/* SPDX-License-Identifier: BSD-3-Clause
```
name the file as just virtio.c

drop the _api.

paravmellanox commented on drivers/vdpa/virtio/virtio.h:
```diff
@@ -0,0 +1,283 @@
+	bool use_va;
+};
+
+struct virtio_ops {
```
is this code copied/modified from the existing dpdk code?
if so, can we use it in its original form?

paravmellanox commented on drivers/vdpa/virtio/virtio.h:
```diff
@@ -0,0 +1,283 @@
+
+#include <rte_ether.h>
+
+/* The feature bitmap for virtio net */
```
please keep net specific items in the virtio_net.h header file out of here.

kailiangz1 committed:
virtio/dev: Add common virtio device driver

The interface of virtio device driver is virtio_api.h All the api function have parameter struct virtio_pci_dev* and it is
allocated in function virtio_dev_alloc

Signed-off-by: Kailiang Zhou <kailiangz@nvidia.com>

kailiangz1 committed:
vdpa/virtio: Add new virtio vdpa driver

This vdpa driver works on top of virtio device layer use virtio_api.h
Only support 1 queue pair for virtio net device.

Signed-off-by: Kailiang Zhou <kailiangz@nvidia.com>

paravmellanox commented on drivers/common/virtio/meson.build:
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright 2019 Mellanox Technologies, Ltd
```
change to latest copy right.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+virtio_dev_alloc(struct rte_pci_device *pci_dev)
+{
+	struct virtio_pci_dev *vpdev;
+	struct virtio_hw *hw = NULL;
```
do not need to do null initialize.
you are assigning it to valid value on line 28

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+#include "virtio_logs.h"
+
+struct virtio_pci_dev *
+virtio_dev_alloc(struct rte_pci_device *pci_dev)
```
I think the api name should be virtio_pci_dev_alloc.

because it returns virtio_pci_dev pointer.

Please also change other friend functions.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+	VTPCI_DEV(hw) = pci_dev;
+	ret = virtio_dev_pci_init(pci_dev, vpdev);
+	if (ret) {
+		PMD_INIT_LOG(ERR, "Failed to init virtio PCI device");
```
better to print the error value along with the pci device name so that we exactly know which device and error occured.

otherwise its unknown to device under debug.

Align other similar prints too.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+		goto error;
+	}
+
+	/* Reset the device although not necessary at startup */
```
better comment is,

Reset the device, so when device is used later, reset time is saved.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+	/* Reset the device although not necessary at startup */
+	virtio_dev_reset(vpdev);
+
+	/* Tell the host we've noticed this device. */
```
This tells to the deice.

so please change comment to

Tell the device that driver has noticed it.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+	/* Tell the host we've noticed this device. */
+	virtio_dev_set_status(vpdev, VIRTIO_CONFIG_STATUS_ACK);
+
+	/* Tell the host we've known how to drive the device. */
```
similar to above comment.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+error:
+		if (vpdev) {
+			rte_free(vpdev);
+		}
```
no need for the extra bracket. please remove.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+	return vpdev;
+
+error:
+		if (vpdev) {
```
In all the cases when you jump to the error, vpdev is valid pointer.
so please remove this check.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+			rte_free(vpdev);
+		}
+		return NULL;
+
```
remove extra line

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+void
+virtio_dev_free(struct virtio_pci_dev *vpdev)
+{
+	struct virtio_hw *hw = NULL;
```
remove this null initializing, and check whole code for similar unwanted code pattern.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+	virtio_dev_set_status(vpdev, VIRTIO_CONFIG_STATUS_DRIVER);
+
+	hw->guest_features = VIRTIO_OPS(hw)->get_features(hw);
+	PMD_INIT_LOG(DEBUG,"Guest_features is 0x%"PRIx64,hw->guest_features);
```
extra white space needed before hw->guest_features.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+
+	hw = &vpdev->hw;
+	if (VIRTIO_OPS(hw)->dev_close(hw))
+		PMD_INIT_LOG(ERR, "Failed to close virtio device.");
```
remove "full stop" = "." after device and in all the logs.
keep it consistent with all the logs.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+		if (!vq)
+			continue;
+		rte_free(vq);
+		hw->vqs[i] = NULL;
```
no need for null initializing the VQ structure itself is getting freed. in line 102.

remove this line.

kailiangz1 replied:
it is needed, vqs is a 2 dimesional array, the fist dimension is array pointer of num_queue.
and it is alloced.
so need this free.

kailiangz1 replied:
It is just in case double free, it shuld be useful.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+}
+
+void
+virtio_dev_free_queues(struct virtio_pci_dev *vpdev,uint16_t nr_vq)
```
extra white space needed before uint16_t

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+	virtio_dev_set_status(vpdev, VIRTIO_CONFIG_STATUS_ACK);
+
+	/* Tell the host we've known how to drive the device. */
+	virtio_dev_set_status(vpdev, VIRTIO_CONFIG_STATUS_DRIVER);
```
we are following module_sub_module_object_verb() terminology.

so
virtio = module
pci_dev = object
status = sub_object
set = verbs
virtio_pci_dev_set_status().

Please audit all the names and change for queues_free() and more.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+	struct virtqueue *vq;
+	uint16_t i;
+
+	hw->vqs = rte_zmalloc(NULL, sizeof(struct virtqueue *) * nr_vq, 0);
```
do sizeof(vq) instead of sizeof(sizeof(struct virtqueue *)).

This ensures that any structure name change etc doesn't affect the code tomorrow.
it will be always size of vq structure regardless of its type.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+	}
+
+	for (i = 0; i < nr_vq; i++) {
+		vq = rte_zmalloc_socket(NULL, sizeof(struct virtqueue), RTE_CACHE_LINE_SIZE,
```
sizeof(*vq)

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+					VTPCI_DEV(hw)->device.numa_node);
+		if (vq == NULL) {
+			PMD_INIT_LOG(ERR, "Failed to alloc virtio queue.");
+			virtio_dev_free_queues(vpdev,nr_vq);
```
white space before nr_vq

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+			virtio_dev_free_queues(vpdev,nr_vq);
+			return -ENOMEM;
+		}
+		hw->vqs[i] = vq;
```
though it doesnt matter much here, before publishing a vq pointer in an array, it is better to initailize this vq structure fully.

So line 127 should come after line 129.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+}
+
+void
+virtio_dev_get_features(struct virtio_pci_dev *vpdev,uint64_t *features)
```
all places you need keep white space after "," in functions declaration.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+}
+
+int
+virtio_dev_get_notify_area(struct virtio_pci_dev *vpdev, int qid, uint64_t *offset, uint64_t *size)
```
this line is crossing 80 characters, trim to two lines.

also qid is uint16_t field.

please change the type in all the code.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+	//TO_DO: check whether notify_addr is 4k aligned
+	*offset = (uint8_t*)hw->vqs[qid]->notify_addr - (uint8_t*)VTPCI_DEV(hw)->mem_resource[vpdev->notify_bar].addr + reg.offset;
+	//nofiy area should 4k aligned, so should not just one doorbell size
+	*size = 0x1000;
```
it should be page size aligned. and not 4K. please change the check for PAGE_SIZE

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+}
+
+int
+virtio_dev_set_queue(struct virtio_pci_dev *vpdev, int index, struct virtio_dev_vring_info *vring_info)
```
virtio_dev_vring_info  should be const pointer as this function is not supposed to touch it.

Trim to two lines.

paravmellanox replied:
keep code consistent when you operate on a vq.

either qid or q_index or vq_index or vq_idx.

one place index and other place qid is not good.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+	VIRTIO_OPS(hw)->intr_detect(hw);
+	if(vpdev->msix_status != VIRTIO_MSIX_ENABLED) {
+		PMD_INIT_LOG(ERR, "Error MSI-X not enabled,status: %d", vpdev->msix_status);
+		return -1;
```
-EINVAL instead of -1.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+	if (VIRTIO_OPS(hw)->set_config_irq(hw, 0) ==
+			VIRTIO_MSI_NO_VECTOR) {
+		PMD_INIT_LOG(ERR, "Failed to set config vector");
+		return -1;
```
see if you can error code from this callback and use that error code instead of -1.

if not, do -EINVAL.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+		}
+	}
+	return 0;
+
```
remove empty line.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+	for (i = 0; i < nr_vring; i++) {
+		if (VIRTIO_OPS(hw)->set_queue_irq(hw, hw->vqs[i], i + RTE_INTR_VEC_RXTX_OFFSET) ==
+						 VIRTIO_MSI_NO_VECTOR) {
+			PMD_INIT_LOG(ERR, "Failed to set queue vector");
```
print the queue number.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+}
+
+int
+virtio_dev_interrupts_disable(struct virtio_pci_dev *vpdev, int nr_vring)
```
disable must be a reverse sequence of enable.

so code of line 250 should be done after the for loop.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,339 @@
+	hw = &vpdev->hw;
+	hw_vq = hw->vqs[index];
+
+	hw_vq->vq_ring_mem  = vring_info->desc;
```
no need for white space before = size, cut it.

paravmellanox commented on drivers/common/virtio/virtio.h:
```diff
@@ -0,0 +1,206 @@
+/*
+ * Maximum number of virtqueues per device.
+ */
+#define VIRTIO_MAX_VIRTQUEUE_PAIRS 8
```
we dont want to depend on max_vq pairs. please remove this and above comment.

paravmellanox commented on drivers/common/virtio/virtio.h:
```diff
@@ -0,0 +1,206 @@
+ * Maximum number of virtqueues per device.
+ */
+#define VIRTIO_MAX_VIRTQUEUE_PAIRS 8
+#define VIRTIO_MAX_VIRTQUEUES (VIRTIO_MAX_VIRTQUEUE_PAIRS * 2 + 1)
```
why +1.
lets remove it.

paravmellanox replied:
we dont want to depends on max vq pairs to get max vqs.

you have nr_vq API.

paravmellanox commented on drivers/common/virtio/virtio.h:
```diff
@@ -0,0 +1,206 @@
+	uint8_t use_inorder_tx;
+	uint8_t opened;
+	uint16_t port_id;
+	uint8_t mac_addr[RTE_ETHER_ADDR_LEN];
```
we dont need mac address at this level.

paravmellanox commented on drivers/common/virtio/virtio.h:
```diff
@@ -0,0 +1,206 @@
+	uint8_t duplex;
+	uint8_t intr_lsc;
+	uint16_t max_mtu;
+	size_t max_rx_pkt_len;
```
many net specific parameters to take this out or put in virtio_net_config new structure.

paravmellanox commented on drivers/common/virtio/virtio.h:
```diff
@@ -0,0 +1,206 @@
+	 * such a contention.
+	 */
+	rte_spinlock_t state_lock;
+	struct rte_mbuf **inject_pkts;
```
no mbufs please.

paravmellanox commented on drivers/common/virtio/virtio.h:
```diff
@@ -0,0 +1,206 @@
+};
+
+#define VIRTIO_OPS(hw)	((hw)->virtio_ops)
+
```
extra white space.

paravmellanox commented on drivers/common/virtio/virtio_pci.c:
```diff
@@ -0,0 +1,829 @@
+
+const struct virtio_ops virtio_dev_pci_legacy_ops = {
+	.read_dev_cfg	= legacy_dev_config_read,
+	.write_dev_cfg	= legacy_dev_config_write,
```
remove extra white spaces

paravmellanox commented on drivers/common/virtio/virtio_pci.c:
```diff
@@ -0,0 +1,829 @@
+io_write64_twopart(uint64_t val, uint32_t *lo, uint32_t *hi)
+{
+	rte_write32(val & ((1ULL << 32) - 1), lo);
+	rte_write32(val >> 32,		     hi);
```
extra white spaces.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.h:
```diff
@@ -0,0 +1,60 @@
+	uint64_t guest_features;
+	pthread_mutex_t vq_config_lock;
+	int configured;
+	uint16_t nr_virtqs;   //number of vq vhost enabled
```
no c++ comments.

do /* */

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,970 @@
+}
+
+static int
+virtio_vdpa_vfio_intr_enable(struct virtio_vdpa_priv *priv)
```
No.

This code belongs in the low level virtio pci driver at virtio_pci_dev level.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,970 @@
+}
+
+static int
+virtio_vdpa_virtq_doorbell_relay_disable(struct virtio_vdpa_priv *priv, int index)
```
we always want to work with direct doorbells.

so please remove this code.

kailiangz1 replied:
no, mlx5_vdpa still have this code and it can be used because some time vhost msg come and 
doorbell map still not run.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,970 @@
+	struct virtio_dev_vring_info vring_info;
+	uint64_t gpa;
+
+	if(enable==0) {
```
white spaces

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,970 @@
+
+	DRV_LOG(INFO, "vDPA device %d  set vring %d state %d.", vid, vring, state);
+	priv->vrings[vring]->enable = !!state;
+	if((priv->configured)&&(state)&&(vring== (rte_vhost_get_vring_num(vid)-1))) {
```
something looks hacky here, or I don;t fully understand this.

kailiangz1 replied:
will move to driver_ok after Yajun's patch, this is temporary solution.

paravmellanox commented on drivers/common/virtio/virtio_ring.h:
```diff
@@ -0,0 +1,161 @@
+
+struct vring {
+	unsigned int num;
+	struct vring_desc  *desc;
```
please dont add any extra white space to right align.

C does left align the code. remove it.

kailiangz1 replied:
done

paravmellanox commented on drivers/common/virtio/virtio_ring.h:
```diff
@@ -0,0 +1,161 @@
+};
+
+struct vring {
+	unsigned int num;
```
have structure nauturally aligned without holds unless it is absolute necessary.

so move this to the end of the struct.

paravmellanox commented on drivers/common/virtio/virtio_ring.h:
```diff
@@ -0,0 +1,161 @@
+};
+
+struct vring_packed {
+	unsigned int num;
```
move this variable to end of the structure.

kailiangz1 replied:
done

paravmellanox commented on drivers/common/virtio/virtio_pci.c:
```diff
@@ -0,0 +1,829 @@
+				break;
+			}
+
+			if (flags & PCI_MSIX_ENABLE)
```
better rewrite it as,

return (flags & PCI_MSIX_ENABLE) ? VIRTIO_MSIX_ENABLED : VIRTIO_MSIX_DISABLED;

you can see such usage in VIRTIO_PCI_CONFIG.

paravmellanox commented on drivers/common/virtio/virtio_pci.c:
```diff
@@ -0,0 +1,829 @@
+ * Copyright (c) 2022 NVIDIA Corporation & Affiliates
+ */
+
+#include <stdint.h>
```
if you are moving the code from X file to Y file.
Do not copy, you have to delete the code from some file that from where you are moving.

paravmellanox commented on drivers/common/virtio/virtio_pci.c:
```diff
@@ -0,0 +1,829 @@
+	return 0;
+}
+
+const struct virtio_ops virtio_dev_pci_modern_ops = {
```
this should be static.

same for the legacy ops too.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,443 @@
+
+	return irq_info.count;
+}
+#define VFIO_FD_INVALID -1
```
empty line needed before this define

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,443 @@
+	}
+	return 0;
+}
+int
```
blank line needed

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,443 @@
+	rte_free(irq_set);
+	if (ret) {
+		PMD_INIT_LOG(ERR, "Dev %s enabling MSI-X: %s", VP_DEV_NAME(vpdev), strerror(errno));
+		return ret;
```
no need for this return inside.

always do return ret on line 260.

paravmellanox commented on drivers/common/virtio/virtio.c:
```diff
@@ -0,0 +1,443 @@
+	struct virtio_hw *hw = &vpdev->hw;
+
+	VIRTIO_OPS(hw)->set_status(hw, VIRTIO_CONFIG_STATUS_RESET);
+	/* Flush status write and wait device ready max 3 seconds. */
```
3 seconds is too less for us during dpu reboots.

it should be close to 120 secs.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.h:
```diff
@@ -0,0 +1,63 @@
+/* SPDX-License-Identifier: BSD-3-Clause
```
since there is single file who is using this, there is no need of a separate header file.

Move them to virtio_vdpa.c

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.h:
```diff
@@ -0,0 +1,63 @@
+	uint64_t used;
+	uint16_t size;
+	uint16_t last_avail_idx;
+	uint16_t last_used_idx; /*other is for future use, not use now*/
```
it should be,

/* It is for future use */

Actually remove it from this patch, if we are not using it.
lets add when you actually use it.
it keeps review and code small.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.h:
```diff
@@ -0,0 +1,63 @@
+	uint16_t last_used_idx; /*other is for future use, not use now*/
+	bool enable;
+	uint8_t notifier_state;
+	uint16_t index;
```
move index variable up.
keep things naturally aligned.
so move bool at the end of the struct along with notifier_state.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.h:
```diff
@@ -0,0 +1,63 @@
+};
+
+#define VIRTIO_VDPA_DRIVER_NAME vdpa_virtio
+//TO_DO: need to add VHOST_USER_PROTOCOL_F_CONFIG later
```
no cpp style comment.
/* Comment */

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.h:
```diff
@@ -0,0 +1,63 @@
+*/
+struct virtio_vdpa_priv {
+	TAILQ_ENTRY(virtio_vdpa_priv) next;
+	struct rte_pci_device *pdev;
```
evaluate why do we need to know about pdev.
It should not be here.
we are working on top of virtio_pci_dev vpdev or rte_vdpa_device.

we should not directly access rte_pci_device at this layer.
Please remove or schedule a call for all the things that I do not understand.

kailiangz1 replied:
we need to access intr_handle at virtio-vdpa.c to configure device level interrupt .
intr_handle  is stored in rte_pci_device

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.h:
```diff
@@ -0,0 +1,63 @@
+	int nvec;
+	uint64_t guest_features;
+	pthread_mutex_t vq_config_lock;
+	int configured;
```
this is likely bool from the name.
if so, please change it to bool

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.h:
```diff
@@ -0,0 +1,63 @@
+	int vid;
+	int nvec;
+	uint64_t guest_features;
+	pthread_mutex_t vq_config_lock;
```
Add comment what is protected by this mutex.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.h:
```diff
@@ -0,0 +1,63 @@
+	uint64_t guest_features;
+	pthread_mutex_t vq_config_lock;
+	int configured;
+	uint16_t nr_virtqs;   /*number of vq vhost enabled*/
```
Please consistency in the code comments all over.

/* First letter to be capital, and have white space at the start and end, no full stop */

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.h:
```diff
@@ -0,0 +1,63 @@
+	struct virtio_vdpa_vring_info **vrings;
+};
+
+#define VIRTIO_VDPA_INTR_RETRIES_USEC 1000
```
also, please schedule meeting to discuss the need of this.
This is strange area.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+#include <rte_log.h>
+#include <rte_errno.h>
+#include <rte_string_fns.h>
+#include <rte_bus_pci.h>
```
mostly we should not need to include this header.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+#include <sys/socket.h>
+#include <sys/ioctl.h>
+#include <fcntl.h>
+#include <netinet/in.h>
```
I dont think this header should be needed here. and also of the line 5.
Please double check all the header file inclusions.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+#include <rte_errno.h>
+#include <rte_string_fns.h>
+#include <rte_bus_pci.h>
+#include <rte_vfio.h>
```
this one too. we dont talk directly to the vfio layer.

kailiangz1 replied:
we need #include <rte_vfio.h>
dma map is done in this layer

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	rte_log(RTE_LOG_ ## level, virtio_vdpa_logtype, \
+		"VIRTIO VDPA %s(): " fmt "\n", __func__, ##args)
+
+TAILQ_HEAD(virtio_vdpa_privs, virtio_vdpa_priv) virtio_priv_list =
```
this should be static as well.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	.remove = virtio_vdpa_dev_remove,
+};
+
+RTE_PMD_REGISTER_PCI(VIRTIO_VDPA_DRIVER_NAME, virtio_vdpa_driver);
```
I think we are the upper layer of this pci device.
I am thinking this code should go to virtio pci driver and not the vdpa driver.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+		struct rte_pci_device *pci_dev)
+{
+	int vdpa = 0;
+	int ret = 0;
```
please no zero intialize it when it is initalize first time in line 796.

This is generic comment. So please check all the line of code again for this code pattern.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+
+	ret = virtio_pci_devargs_parse(pci_dev->device.devargs, &vdpa);
+	if (ret < 0) {
+		DRV_LOG(ERR, "Devargs parsing is failed");
```
print the error code ret.

Any error print that doesnt prin the error and device is not of much use.
Evaluate all the code and prints that is missing error code and device id/name in it

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+		return ret;
+	}
+	/* virtio vdpa pmd skips probe if device needs to work in none vdpa mode */
+	if (vdpa != 1)
```
this seems boolean as true/false for the vdpa.

So please change it from it to bool

kailiangz1 replied:
this is a int value. not bool. we can add more mode if needed

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	}
+	/* virtio vdpa pmd skips probe if device needs to work in none vdpa mode */
+	if (vdpa != 1)
+		return 1;
```
return -EINVAL

kailiangz1 replied:
return 1 is not means wrong. it means customer don't want vdpa driver work in vdpa mode,
should return 1 let bus continure probe

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	if (vdpa != 1)
+		return 1;
+
+	priv = rte_zmalloc("virtio vdpa device private", sizeof(*priv), RTE_CACHE_LINE_SIZE);
```
first time allocation, no need for NULL init on line 792.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+
+	priv->pdev = pci_dev;
+
+	priv->vpdev = virtio_pci_dev_alloc(pci_dev);
```
code of line 843 to 855 and code of line 864 to 879 belongs to the virtio_pci driver.

He is supposed to initialize this low level details. To that layer pass only the pci device and it will take care of it.

kailiangz1 replied:
vdpa driver need intr_handle to handel software doorbell. it is needed in this layer.
also, it is needed by vdpa driver to know how many interrupt device supported.

kailiangz1 replied:
we need to know how many interrupts device support to judge 
whether to fail when queue is bigger.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+			break;
+		}
+	}
+	if (found)
```
no need for this check here, move this line to line 909.

kailiangz1 replied:
it is protected by lock, shuld keeps as it now.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	if (found)
+		TAILQ_REMOVE(&virtio_priv_list, priv, next);
+	pthread_mutex_unlock(&priv_list_lock);
+	if (found) {
```
write this as,

if (!found)
  return -ENODEV:

Code of line 917 to .. comes here.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+		if (priv->configured)
+			virtio_vdpa_dev_close(priv->vid);
+
+		if (priv->vdev)
```
it is not possible to call this function with vdev is null.
remove this check.
also when you move this code to virtio layer, this check is not necessary.

kailiangz1 replied:
what about probe have error and vdpa is not registered

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	}
+
+	//TO_DO: need to confirm following:
+	priv->vfio_dev_fd = -1;
```
I think all of these should stay inside the virtio driver operating over vfio.
we at this driver layer only use it.

kailiangz1 replied:
DMA map operation still needs this

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	}
+
+	priv->hw_nr_virtqs = virtio_pci_dev_nr_vq_get(priv->vpdev);
+	ret = virtio_vdpa_queues_alloc(priv);
```
this code might stay here.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+		}
+		priv->vrings[i] = vr;
+		priv->vrings[i]->index = i;
+		priv->vrings[i]->priv = priv;
```
lets double check if for all 256 vqs do we need to store this pointer or we can get the priv from the device itself.

kailiangz1 replied:
for each vq, we need store a fd to hanle softare doorbell.
we need to rember enable state in vq, to disable vq first if vq is enabled in cause of modify

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+		priv->vrings = NULL;
+	}
+
+	virtio_pci_dev_queues_free(priv->vpdev,nr_vq);
```
white space before nr_vq

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	struct virtio_vdpa_vring_info *vr;
+	uint16_t i;
+
+	if(priv->vrings) {
```
this check is not needed.
if the queues are not allocated, device registration fails.

remove it.

kailiangz1 replied:
it is needed just in case upper code has error, we didn't free null pointer

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	struct vfio_region_info reg = { .argsz = sizeof(reg) };
+	int ret;
+
+	if (priv == NULL) {
```
when we come here for the registered device, priv cannot be done.

remove this check. In this routine and all other routines.

kailiangz1 replied:
Just in cause, code always have bugs

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	struct rte_vdpa_device *vdev = rte_vhost_get_vdpa_device(vid);
+	struct virtio_vdpa_priv *priv =
+		virtio_vdpa_find_priv_resource_by_vdev(vdev);
+	struct vfio_region_info reg = { .argsz = sizeof(reg) };
```
this functionality of vfio stays in the lower layer.

Please move and provide an API.

kailiangz1 replied:
vfio_region_info removed

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+		return -ENODEV;
+	}
+	if (priv->configured) {
+		DRV_LOG(ERR, "Failed to reconfigure vid %d.", vid);
```
print should be, "Device already configured,

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	}
+	if (priv->configured) {
+		DRV_LOG(ERR, "Failed to reconfigure vid %d.", vid);
+		return -EINVAL;
```
instead of EINVAL, it should be EBUSY

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	if (priv->nvec <= (priv->nr_virtqs +1)) {
+		DRV_LOG(ERR, "Error dev interrupts %d less than queue: %d",
+			priv->nvec, priv->nr_virtqs +1);
+		return -EINVAL;
```
you left the device in half configured state with configured = 1 on line 591, and dma map already created,

you need to unwind the error by having jump label goto to do dma_unmap()

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	if (ret) {
+		DRV_LOG(ERR, "Error enabling virtio dev interrupts: %s",
+				strerror(errno));
+		return ret;
```
have jump label and unwind the error.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	priv->configured = 1;
+	DRV_LOG(INFO, "vDPA device %d was configured.", vid);
+
+	fd =rte_intr_fd_get(priv->pdev->intr_handle);
```
white space.

You should run kernel's checkpatch.pl to catch these silly cosmetic errors.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+		return ret;
+	}
+	priv->configured = 1;
+	DRV_LOG(INFO, "vDPA device %d was configured.", vid);
```
no full stop at the end of the log, similar to line 588

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	}
+
+	priv->nr_virtqs = rte_vhost_get_vring_num(vid);
+
```
remove this blank line, not needed.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+
+	/* The mutex may stay locked after event thread cancel - initiate it. */
+	pthread_mutex_init(&priv->vq_config_lock, NULL);
+	DRV_LOG(INFO, "vDPA device %d was closed.", vid);
```
no full stop

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+			reg->host_user_addr, reg->guest_phys_addr,
+			reg->size);
+		if (ret < 0) {
+			DRV_LOG(ERR, "DMA map failed.");
```
error code here along with device vid

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+			reg->size);
+		if (ret < 0) {
+			DRV_LOG(ERR, "DMA map failed.");
+			goto exit;
```
you have partially mapped and partially unmap.

add error unwinding code to do dma_unmap on when it fails half way.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+		return -E2BIG;
+	}
+
+	//TO_DO: check if vid set here is sutiable
```
suitable

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	}
+
+	//TO_DO: check if vid set here is sutiable
+	priv->vid = vid;
```
Do

yum install -y codespell
git format-patch -2 -o /tmp/patches/
codespell /tmp/patches/*

This will emit all the spelling errors and fix them.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	if (state && !priv->vrings[vring]->enable) {
+		ret = virtio_vdpa_virtq_enable(priv, vring);
+	}
+	if (state && priv->vrings[vring]->enable) {
```
when state = 1, and if the ring is already enabled, why do you disable it?

i thin it should be,

if (!state && priv->vring->enable) 
you should disable it.

kailiangz1 replied:
state=1 means parameter change, so need to disable, then enable to new parameter

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+			pthread_mutex_unlock(&priv->vq_config_lock);
+			return ret;
+		}
+		ret = virtio_vdpa_virtq_enable(priv, vring);
```
from this line, it looks like we never disable the vq.

kailiangz1 replied:
in 	if (!state && priv->vrings[vring]->enable) {
we disable vq

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+}
+
+static int
+virtio_vdpa_vring_state_set(int vid, int vring, int state)
```
please keep the code consistent,

either vring_index or vring or just index

here you say vring and on line 280, you say index.
so please change everything to vq_idx.
it indicates that its an virtiq index.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	if (ret)
+		return ret;
+
+	ret =virtio_pci_dev_interrupts_enable(priv->vpdev, vq.callfd, index+1);
```
white space after =

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	if (ret)
+		return ret;
+
+	ret =virtio_pci_dev_interrupts_enable(priv->vpdev, vq.callfd, index+1);
```
this should be done only one, or its name should be pci_dev_interrupt_enable

kailiangz1 replied:
interrupt is enabled per queue, so, i just need to change to pci_dev_interrupt_enable?

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	priv->vrings[index]->size = vq.size;
+	vring_info.size = vq.size;
+
+	DRV_LOG(DEBUG, "Virtq %d nr_entrys:%d",index, vq.size);
```
white space before index

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+
+	priv->vrings[index]->notifier_state = VIRTIO_VDPA_NOTIFIER_STATE_DISABLED;
+
+	ret = virtio_pci_dev_interrupts_disable(priv->vpdev, index+1);
```
white space 

index + 1

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+
+	virtio_pci_dev_queue_del(priv->vpdev, index);
+
+	priv->vrings[index]->notifier_state = VIRTIO_VDPA_NOTIFIER_STATE_DISABLED;
```
this should be after interrupt disablement call is successful.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,955 @@
+	uint64_t buf;
+	int nbytes;
+
+	if(!(priv->configured))
```
braces before priv is not needed.

it should be, 

if (!priv->configured || !virtq->enable)
  return;

cut down line 142 and 143.

paravmellanox commented on drivers/common/virtio/virtio_pci.c:
```diff
@@ -0,0 +1,823 @@
+ */
+
+#ifdef RTE_EXEC_ENV_LINUX
+ #include <dirent.h>
```
no white space needed at beginning.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,1045 @@
+virtio_vdpa_find_priv_resource_by_vdev(const struct rte_vdpa_device *vdev)
+{
+	struct virtio_vdpa_priv *priv;
+	int found = 0;
```
use bool 

bool found = false;

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,1045 @@
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
+	}
+	/* TO_DO:for net, should use max_virtqueue_pairs? may be is same */
```
remove this todo, it belong to virtio layer.

virtio layer will do max_qps vs other count.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,1045 @@
+
+static int
+virtio_vdpa_virtq_doorbell_relay_enable(struct virtio_vdpa_priv *priv,
+															int vq_idx)
```
fix alignment.
cannot figure out from this patch if its right or not.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,1045 @@
+	}
+
+	return 0;
+error:
```
empty line needed before the label.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,1045 @@
+
+	if (virtio_pci_dev_get_status(priv->vpdev) &
+		VIRTIO_CONFIG_STATUS_DRIVER_OK) {
+		DRV_LOG(ERR, "Can not set vring stat when driver ok vDPA device: %s",
```
comment should be

vring state 

not 
vring "stat" -> stat is  usually for the statistics.
please add "e".

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,1045 @@
+	if (!state && priv->vrings[vq_idx]->enable)
+		ret = virtio_vdpa_virtq_disable(priv, vq_idx);
+
+	if (state && !priv->vrings[vq_idx]->enable)
```
change to 
if()
else if ()
else if ()


because all conditions cannot be true at same time.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,1045 @@
+}
+
+static int
+virtio_vdpa_vring_state_set(int vid, int vq_idx, int state)
```
change the variable name from "state" to "enable".

you can see in the below 3 conditions that enable reads the code better and aligns to the enable/disable vq calls that we are making there.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,1045 @@
+		return ret;
+	}
+
+	priv->vrings[vq_idx]->enable = 1;
```
change this enable data type to be bool and do true/false

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,1045 @@
+	if (state && !priv->vrings[vq_idx]->enable)
+		ret = virtio_vdpa_virtq_enable(priv, vq_idx);
+
+	if (state && priv->vrings[vq_idx]->enable) {
```
virtio_vdpa_virtq_enable() and virtio_vdpa_virtq_disable() functions are changing the value of this enable/disable flag inside.

So without if-else, this logic is hard to see, unless you expliclty wanted that way.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,1045 @@
+		ret = virtio_vdpa_virtq_enable(priv, vq_idx);
+
+	if (state && priv->vrings[vq_idx]->enable) {
+		ret = virtio_vdpa_virtq_disable(priv, vq_idx);
```
also this disable/enable is tricky.

Please add a comment explaining why is this done.
Why a queue that is already working is disabled and enabled again? Why no-op is not enough on already enabled vq.
And why someone calls queue enable on already enabled vq.
Please add comment block explaining this wierdness.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,1045 @@
+					priv->vdev->device->name, ret);
+	}
+
+	virtio_pci_dev_reset(priv->vpdev);
```
This needs to be done before doing dma unmap.
otherwise memory is unmapped but device VQ are still working on these vqs.

So reset the device so that VQ traffic stops.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,1045 @@
+
+	virtio_pci_dev_reset(priv->vpdev);
+
+	/* Disable all queue*/
```
add white space after "queue"

paravmellanox replied:
and say "queues".

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,1045 @@
+	}
+
+	virtio_pci_dev_queues_free(priv->vpdev, nr_vq);
+
```
remove this empty line.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,1045 @@
+	struct virtio_vdpa_priv *priv = NULL;
+	int found = 0, ret;
+
+	pthread_mutex_lock(&priv_list_lock);
```
use the function you have start the beginning of the file.

paravmellanox commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -0,0 +1,1045 @@
+		rte_free(priv);
+	}
+
+	if (!found)
```
return found ? 0 : -ENODEV;
